### ISSUE TITLE
Podman failing to pull a missing image, gives the user a horrible error.

### ISSUE THREAD

**Issue by rhatdan** [containers/podman#9619]
```
# podman pull  docker.io/library/foobar:latest
Trying to pull docker.io/library/foobar:latest...
  denied: requested access to the resource is denied
Error: Error initializing source docker://foobar:latest: Error reading manifest latest in docker.io/library/foobar: errors:
denied: requested access to the resource is denied
unauthorized: authentication required
```

Versus

```
# docker  pull  docker.io/library/foobar:latest
Error response from daemon: pull access denied for foobar, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

The issue we have is that podman can not easily translate the errors.Cause coming back from podman to display this error.

**Comment by rhatdan** [comment]
@vrothberg Could we change the containers/image to return us a constant error that we could compare against, so that we could put out a decent error message.

https://github.com/containers/image/blob/0495207afcac19a76cc723085abe56ba393d0cd5/docker/errors.go#L39

**Comment by rhatdan** [comment]
@mtrmac WDYT?

**Comment by vrothberg** [comment]
I think that we can tackle it in c/image and stop from wrapping the error too often. The critical part is that Docker explains that the resource may not exist.  That's something the error from the registry does not and is really confusing.

**Comment by mtrmac** [comment]
Make sure to dump the full error object and attribute where the various parts come from; IIRC some of the text comes from the server side, and it’s awkward to just throw away server-side strings (as the only error information that might actually be available to clients that don’t operate the registry). OTOH it might make sense to specifically match and edit that particular server-side string if it is constant and consistently unhelpful.

A few years ago Brent looked into making the “image not found or privileges insufficient” behavior consistent among the major registries, and just gave up because they were failing in so many different ways. A few years later might be a good time to revisit if things have improved — just expect a mess.

**Comment by rhatdan** [comment]
What if we defined a constant error and then wrapped it.
	ErrAuthorozationFailed = errors.New("repository does not exist or may require credentials")

	return errors.Wrap(ErrTooManyRequests, ErrUnauthorizedForCredentials{Err: err})

        Now in podman we can look for errors.Cause(err)=0ErrTooManyRequests and report that.

**Comment by mtrmac** [comment]
Sure. The work is in defining the “it” to wrap, without discarding things that should be preserved.

- IIRC one of the major registries used to just return a 404 HTTP status with fairly unhelpful HTML content. Probably should be wrapped.
- What if the server returns the same correctly-parsed error codes, but instead of the generic “requested access to the resource is denied” it says “you have been rate-limited, visit $personalized-URL to unlock”? That should not just be thrown away.

**Comment by rhatdan** [comment]
Well should we just look for a string match on the most common error?  If we match we put out something helpful, if not we throw up on the screen as we do now.

It would be nice if Containers/image could tell us if this is the common error.

**Comment by mtrmac** [comment]
Sure, c/image _should_ do all of that, and provide a well-typed error (… with _some_ semantics; what _exactly_ would that be? “(the image does not exist with that name, _or_ is inaccessible with provided credentials) _and_ there is no more information available”?.

I’m just cautioning not to expect this to be an obvious one-liner.

**Comment by rhatdan** [comment]
That sounds good to me.  If I printed out the error.

the image does not exist with that name, or is inaccessible with provided credentials
The caller would have a much better experience.

**Comment by rhatdan** [comment]
@lsm5 PTAL

**Comment by rhatdan** [comment]
@lsm5 do you have time to tackle this?

**Comment by vrothberg** [comment]
I am going to merge this with https://github.com/containers/podman/issues/10057.

**Comment by vrothberg** [comment]
To nail down the scope a bit further.  It depends which image we're attempting to pull from which registry.

**Docker**
```
root@ubuntu:/home/vrothberg# docker pull registry.fedoraproject.org/foobar
Using default tag: latest
Error response from daemon: manifest for registry.fedoraproject.org/foobar:latest not found: manifest unknown: manifest unknown
root@ubuntu:/home/vrothberg# docker pull docker.io/library/foobar
Using default tag: latest
Error response from daemon: pull access denied for foobar, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**Podman**
```
~ $ podman pull registry.fedoraproject.org/foobar
Trying to pull registry.fedoraproject.org/foobar:latest...
  manifest unknown: manifest unknown
Error: Error initializing source docker://registry.fedoraproject.org/foobar:latest: Error reading manifest latest in registry.fedoraproject.org/foobar: manifest unknown: manifest unknown
~ $ podman pull docker.io/library/foobar
Trying to pull docker.io/library/foobar:latest...
  denied: requested access to the resource is denied
Error: Error initializing source docker://foobar:latest: Error reading manifest latest in docker.io/library/foobar: errors:
denied: requested access to the resource is denied
unauthorized: authentication required

```

While the "manifest unknown: manifest unknown" error is the same, the error reporting in Podman stutters a bit with the reference of the image being printed twice. I think we should remove the stuttering (must be done in c/image).

The authorized error is not yet entirely clear to me.  Podman gets a `401` (unauthorized), while Docker seems to get a `403` (forbidden) (see https://github.com/moby/moby/blob/master/distribution/errors.go#L61).

I am not versed enough in the distribution parts to explain this difference. @mtrmac do you know?

**Comment by mtrmac** [comment]
(Also copied from #10057:)

> ```
> ~ $ podman pull registry.fedoraproject.org/foobar
> Trying to pull registry.fedoraproject.org/foobar:latest...
>   manifest unknown: manifest unknown
> Error: Error initializing source docker://registry.fedoraproject.org/foobar:latest: Error reading manifest latest in registry.fedoraproject.org/foobar: manifest unknown: manifest unknown
> ~ $ podman pull docker.io/library/foobar
> Trying to pull docker.io/library/foobar:latest...
>   denied: requested access to the resource is denied
> Error: Error initializing source docker://foobar:latest: Error reading manifest latest in docker.io/library/foobar: errors:
> denied: requested access to the resource is denied
> unauthorized: authentication required
> ```
> 
> While the "manifest unknown: manifest unknown" error is the same, the error reporting in Podman stutters a bit with the reference of the image being printed twice. I think we should remove the stuttering (must be done in c/image).

That sounds unlikely; AFAIK c/image’s copy never writes errors to any output, only progress information. This is almost certainly https://github.com/containers/podman/blob/5b2585f5e91ca148f068cefa647c23f8b1ade622/libpod/image/pull.go#L351 (note the two-space prefix added by `cleanErrorMessage`).

It seems that’s already been removed in libimage.

--

> The authorized error is not yet entirely clear to me. Podman gets a `401` (unauthorized), while Docker seems to get a `403` (forbidden) (see https://github.com/moby/moby/blob/master/distribution/errors.go#L61).

Compare https://github.com/containers/image/issues/1293#issuecomment-875791392 : docker/distribution returns a multi-error, and Docker only shows the first value. Maybe that’s something c/image could do, let’s discuss in that issue.

**Comment by mtrmac** [comment]
And a self-reply…

> > While the "manifest unknown: manifest unknown" error is the same, the error reporting in Podman stutters a bit with the reference of the image being printed twice. I think we should remove the stuttering (must be done in c/image).
> 
> That sounds unlikely; AFAIK c/image’s copy never writes errors to any output, only progress information.

Ah, the original comment was talking about the _name_, while I was looking at the _error_. The trouble with that is that `copy.Image` must let the user know whether it’s the source or the destination that was failing, and “error reading manifest latest in $name” actually reports a specific mirror.

Compare https://github.com/containers/image/pull/1240 , which has also changed the text of the errors a bit.

It might make sense for `copy.Image` to wrap error with `copy.SourceError` / `copy.DestinationError` or something like that, for the caller to strip, I’m not immediately sure how exactly that would work with `pkg/errors.Cause` vs. `errors.As` and existing callers.

**Comment by rhatdan** [comment]
```
 ./bin/podman pull registry.fedoraproject.org/foobar
Trying to pull registry.fedoraproject.org/foobar:latest...
Error: Error initializing source docker://registry.fedoraproject.org/foobar:latest: Error reading manifest latest in registry.fedoraproject.org/foobar: manifest unknown: manifest unknown
```
Looking at this error, I see three stutters:

1 Error: Error
2. initializing source docker://registry.fedoraproject.org/foobar:latest: Error reading manifest latest in registry.fedoraproject.org/foobar: manifest 
3. manifest unknown: manifest unknown

Number `1` Is fixed in C/I but not merged into Podman yet.
Should `2` be fixed in Podman then?  Just don't add the `initializing source ...` wrap?
Number `3` is buried down in docker code, I believe, how should we handle that?

**Comment by mtrmac** [comment]
2 “error initializing source” is added by `copy.Image`. I guess Podman could strip that part if it were API-visible if it is already adding “Trying to pull”, but that’s not _that_ much of a recognizable special case; “initializing source” may succeed but subsequent I/O may fail, and it’s generally necessary to report _what_ failed.

2 “Initializing source $repo:$tag” vs. “reading manifest $tag in $mirror” is not a stutter, that’s relevant information, discussed in https://github.com/containers/image/pull/1240 .

3 should be easy enough to hide with an `Cause`/`Unwrap`-transparent wrapper (detect an `errcode.Error` where `e.Message == e.Code().Message()`, or maybe where `e.Error()` is exactly the expected stereotypical value, to be safer against changes; and re-format it as desired, probably to include only an `e.Code().Message()` and not `e.Code().Value()`. A plausible small change to `c/image/docker/errors.go`; it is would very clearly be not doing any harm _right now_, OTOH there’s a RFE https://github.com/containers/image/issues/1234 that would make this somewhat more difficult (but then again, arguably, once we get into the business of tweaking docker/distribution/registry/client errors, there’s not much reason to stop at any _specific_ amount of tweaking).

**Comment by rhatdan** [comment]
I don't see the inializing source ... versus reading manifest to be relevant to the user.  Either one would tell the user enough to know that the image failed to pull.  I would rather bury the information at debug level then to show something ugly like this to the user.

```
# docker pull registry.fedoraproject.org/foobar
Using default tag: latest
Error response from daemon: manifest for registry.fedoraproject.org/foobar:latest not found: manifest unknown: manifest unknown
# podman pull registry.fedoraproject.org/foobar
Trying to pull registry.fedoraproject.org/foobar:latest...
Error: Error initializing source docker://registry.fedoraproject.org/foobar:latest: Error reading manifest latest in registry.fedoraproject.org/foobar: manifest unknown: manifest unknown
```

**Comment by umohnani8** [comment]
Checking to see if this can be closed @vrothberg @mtrmac @rhatdan

**Comment by rhatdan** [comment]
We have improved 
```
 ./bin/podman pull  docker.io/library/foobar:latest
Trying to pull docker.io/library/foobar:latest...
Error: initializing source docker://foobar:latest: reading manifest latest in docker.io/library/foobar: errors:
denied: requested access to the resource is denied
unauthorized: authentication required
```
 But not as good as
```
# docker  pull  docker.io/library/foobar:latest
Error response from daemon: pull access denied for foobar, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**Comment by baude** [comment]
Well, that is only because you are using docker.io.  Here is docker with quay:

```
$ sudo docker pull quay.io/homer/simpson
Using default tag: latest
Error response from daemon: unauthorized: access to the requested resource is not authorized
```

This is basically the point that @mtrmac makes.  The various registries have no standard like a 404 that we can depend on. So, unless you want to make some custom error message string handler for *every* image registry out there, we are stuck.

**Comment by vrothberg** [comment]
We could do it for the most common registries.  docker.io and quay.io could be easy targets.

**Comment by umohnani8** [comment]
I can add custom error messages for docker.io and quay.io. Do we also want to return the custom error message for when the user uses short names and selects either docker.io or quay.io from the list? Or should we only limit it to when the user enters a fully qualified image name in the pull command?

**Comment by rhatdan** [comment]
We want it in either case.

As a human being, and not someone who knows what the hell a manifest list is.
```
 $ ./bin/podman pull  quay.io/rhatdan/foobar:latest
Trying to pull quay.io/rhatdan/foobar:latest...
Error: initializing source docker://quay.io/rhatdan/foobar:latest: reading manifest latest in quay.io/rhatdan/foobar: unauthorized: access to the requested resource is not authorized
```
This error tells my brain that I need to login. So I do.

```
$ podman login quay.io
Username: rhatdan
Password: 
Login Succeeded!
$ ./bin/podman pull  quay.io/rhatdan/foobar:latest
Trying to pull quay.io/rhatdan/foobar:latest...
Error: initializing source docker://quay.io/rhatdan/foobar:latest: reading manifest latest in quay.io/rhatdan/foobar: name unknown: repository name not known to registry
```
That is better, but I still have no clue what this `manifest` thing is, I was pulling an image. You forced me to login, just to tell me I had a typo.  The Docker error at least gives me a clue earlier in the processes.

Let's look at a couple of other examples.
```
$ ./bin/podman pull  quay.io/library/foobar:latest
Trying to pull quay.io/library/foobar:latest...
Error: initializing source docker://quay.io/library/foobar:latest: reading manifest latest in quay.io/library/foobar: unauthorized: access to the requested resource is not authorized
$ ./bin/podman pull  quay.io/foobar:latest
Trying to pull quay.io/foobar:latest...
Error: initializing source docker://quay.io/foobar:latest: reading manifest latest in quay.io/foobar: StatusCode: 404, <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final/...
```
Can't we just look for the `resource is not authorized` and give a resonable guess that either the image does not exist or you are not logged in to pull it.

Not sure this is much different then other sites as far as the `requested resource is not authorized` error.

**Comment by mtrmac** [comment]
Some work on this has already happened in https://github.com/containers/image/pull/1299 (one error instead of two for `docker.io`, no repeated `manifest unknown: manifest unknown`), but it has uncovered to a few rat holes that I need to resolve first before finishing that work.

**Comment by mtrmac** [comment]
Oh, also: _Whatever_ anyone does WRT heuristic matching, please capture the exact server-side response and turn that into a unit test.

One of the problems with https://github.com/containers/image/pull/1299 is that we now have workarounds for registry errors in `skopeo inspect` that we’ll probably just break because we don’t know what exactly they were working around.

**Comment by dezza** [comment]
I can reproduce this error by pushing a v2s2 image to a private registry with basic auth (authenticated)

```
podman build --format=docker -t reg.example.com/tor ~/src/containers/tor/
```

```
podman push reg.example.com/tor reg.example.com/tor
```

Now trying to run it.

```
podman run reg.example.com/tor
```

```
Trying to pull oci.recv.ws/thelounge:latest...
Error: initializing source docker://oci.recv.ws/thelounge:latest: reading manifest latest in reg.example.com/tor: unauthorized: authentication required
```

Using [snyk](https://docs.snyk.io/products/snyk-container/snyk-cli-for-container-security/advanced-snyk-container-cli-usage#testing-multi-platform-images) for container testing understands this new v2s2 format, but it seems authenticated private registries with podman does not when pulling?

**Comment by rhatdan** [comment]
Still waiting on container/image PR.

**Comment by IvantheDugtrio** [comment]
Can confirm I'm getting the same issue pulling containers from docker.io. Using podman 3.2.1 on Ubuntu 21.10, installed from apt. 
`podman pull docker.io/library/grafana/grafana:5.4.3
Trying to pull docker.io/library/grafana/grafana:5.4.3...
Error: Error initializing source docker://library/grafana/grafana:5.4.3: Error reading manifest 5.4.3 in docker.io/library/grafana/grafana: errors:
denied: requested access to the resource is denied
unauthorized: authentication required
`
When using the podman login command, I am supposed to use my dockerhub login credentials correct, even when downloading containers that are not my own? It has been a while since I used docker so I don't recall needing to use a login when pulling public containers from dockerhub.

**Comment by IvantheDugtrio** [comment]
Nevermind I incorrectly assumed the address for all dockerhub projects had to be docker.io/library/USERNAME/REPO:VERSION.
using docker.io/grafana/grafana:latest worked as expected.

**Comment by abergmeier** [comment]
> One of the problems with containers/image#1299 is that we now have workarounds for registry errors in skopeo inspect that we’ll probably just break because we don’t know what exactly they were working around.

Could you point me to where the workaround is?
I am currently testing nonexistence using `strings.HasSuffix(err.Error(), ": manifest unknown")`.

**Comment by mtrmac** [comment]
I was thinking of https://github.com/containers/skopeo/blob/e422e44fca37604e603fd72c397ed469cd15a735/cmd/skopeo/inspect.go#L204-L209 in particular as an example of lack of foresight to document the details when writing the code, but any code like the above could easily break in any change to the error strings.

**Comment by abergmeier** [comment]
> I was thinking of https://github.com/containers/skopeo/blob/e422e44fca37604e603fd72c397ed469cd15a735/cmd/skopeo/inspect.go#L204-L209 in particular as an example of lack of foresight to document the details when writing the code

Thanks for the pointer. Interesting.
From what I understand (please correct me if wrong) the situation is:

- We have multiple implementations of Container Registries that all handle slightly different in various ways (_quirks_).
- Of all the different package layers (docker*, container*, buildah, skopeo, podman) not one sets out to encapsulate all _quirks_.

**Comment by mtrmac** [comment]
My concern is not the existence of quirks, and not _that_ much even the decentralization, but primarily maintainability: https://github.com/containers/podman/issues/9619#issuecomment-932807773 . Centralization might help with maintainability but isn’t sufficient.

**Comment by rhatdan** [comment]
@vrothberg @umohnani8 @mtrmac This issue seems to have died out.  Did anyone ever make any progress?

**Comment by vrothberg** [comment]
Not to my knowledge. To make progress, I think we either need to prioritize or have regular "issue weeks".

**Comment by rhatdan** [comment]
Well we have an issue week coming up...

**Comment by vrothberg** [comment]
Need to add FIXMEs in the code :^)

**Comment by mtrmac** [comment]
Current status of https://github.com/containers/image/pull/1299 (old/new):
```console
% c --insecure-policy copy docker://docker.io/library/foobar:latest dir:t
FATA[0001] initializing source docker://foobar:latest: reading manifest latest in docker.io/library/foobar: errors:
denied: requested access to the resource is denied
unauthorized: authentication required 

FATA[0002] initializing source docker://foobar:latest: reading manifest latest in docker.io/library/foobar: requested access to the resource is denied 
```

**Comment by mtrmac** [comment]
#15662 will change the message to match the previous comment.

**Comment by rhatdan** [comment]
Closing this ancient issue.
